File: dotnet/Lexy/Lexy.Compiler/Language/Functions/FunctionResults.cs
using System.Collections.Generic;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Functions
{
    public class FunctionResults : ParsableNode
    {
        public IList<VariableDefinition> Variables { get; } = new List<VariableDefinition>();

        public FunctionResults(SourceReference reference) : base(reference)
        {
        }

        public override IParsableNode Parse(IParserContext context)
        {
            var variableDefinition = VariableDefinition.Parse(VariableSource.Results, context);
            if (variableDefinition == null) return this;

            if (variableDefinition.DefaultExpression != null)
            {
                context.Logger.Fail(Reference,
                    $"Result variable '{variableDefinition.Name}' should not have a default value.");
                return this;
            }

            Variables.Add(variableDefinition);

            return this;
        }

        public override IEnumerable<INode> GetChildren() => Variables;

        protected override void Validate(IValidationContext context)
        {
        }
    }
}

File: typescipt/src/language/functions/functionResults.ts


export class FunctionResults extends ParsableNode {
   public Array<VariableDefinition> Variables = list<VariableDefinition>(): new;

   public FunctionResults(SourceReference reference) : base(reference) {
   }

   public override parse(context: IParseLineContext): IParsableNode {
     let variableDefinition = VariableDefinition.Parse(VariableSource.Results, context);
     if (variableDefinition == null) return this;

     if (variableDefinition.DefaultExpression != null) {
       context.Logger.Fail(Reference,
         $`Result variable '{variableDefinition.Name}' should not have a default value.`);
       return this;
     }

     Variables.Add(variableDefinition);

     return this;
   }

   public override getChildren(): Array<INode> {
     return Variables;
   }

   protected override validate(context: IValidationContext): void {
   }
}
